scripts/stabilization.py

```diff
@@ -0,0 +1,207 @@
+    boost::multi_array<double, 2> coords;
+    element->tabulate_dof_coordinates(coords, vertex_coords, cell);
+
+	
+    dolfin_assert(J->value_size() == J_->value_size());
+    dolfin_assert(J_->value_size() == J_n->value_size());
+	
+	//
+    const unsigned int num_nodes = coords.num_elements();
+    const unsigned int n = J->value_size();
+	
+	
+    std::vector<double> basis_derivatives(dim * num_nodes * num_nodes);
+    std::vector<double> derivatives(num_nodes);
+    
+    double *basis_derivatives_pointer;
+    
+    // buffer arrays for function values
+    Array<double> vals_J(n);
+    Array<double> vals_J_(n);
+    Array<double> vals_J_n(n);
+    
+    Array<double> vals_Enorm(1);
+    Array<double> val_dt(1);
+    //Array<double> val_c_e(1);
+    //Array<double> val_nu_max(1);
+    
+    // aggregated function values
+    Array<double> vals_J_all(n * num_nodes);
+    Array<double> vals_J__all(n * num_nodes);
+    Array<double> vals_J_n_all(n * num_nodes);
+
+	// evaluate normalized entropy variation
+	E_norm->eval(vals_Enorm, x, c);
+	dt->eval(val_dt, x, c);
+	//c_e->eval(val_c_e, x, c);
+	//nu_max->eval(val_nu_max, x, c);
+
+
+    double Dh = 0;
+    double max_Dh = 0;
+    
+   	// 
+    std::vector<double> x_(dim);
+    Array<double> x_function(dim);
+
+	// pointers to (vertex) coordinates for evaluation of basis derivatives
+    double* x_basis = &x_[0];
+    const double *vertex_coords_basis = &vertex_coords[0];
+    
+    // index
+    int j = 0;
+
+    // Different for higher dimension, since all coordinates need to be updated
+    for(auto i = coords.origin(); i < (coords.origin() + coords.num_elements()); ++i)  
+    {
+    	for (unsigned int d=0; d<dim; ++d) {
+	    	x_[d] = i[d];
+    	    x_function[d] = i[d];
+        }
+        
+        // evaluate functions at current dof coordinate
+        J->eval(vals_J, x_function, c);
+        J_->eval(vals_J_, x_function, c);
+        J_n->eval(vals_J_n, x_function, c);
+        
+        // aggregate values
+        for (unsigned int e=0; e<dim; ++e) {
+			vals_J_all[j * dim + e] = vals_J[e];
+			vals_J__all[j * dim + e] = vals_J_[e];
+			vals_J_n_all[j * dim + e] = vals_J_n[e];        	
+        }
+        
+        // pointer to current basis derivatives and 
+        basis_derivatives_pointer = &basis_derivatives[num_nodes * j];
+		element->evaluate_basis_derivatives_all(1, basis_derivatives_pointer, x_basis, vertex_coords_basis, (int) cell.orientation());
+
+        j++;
+    }
+    
+     for(int k=0; k < num_nodes * num_nodes; ++k) {
+		derivatives[k / num_nodes] = derivatives[k / num_nodes] + 0.5 * basis_derivatives[k] * pow(vals_J_all[k % num_nodes], 2);
+		
+		if ((k % num_nodes) == num_nodes - 1) {
+			Dh = 0.25 * (3 * pow(vals_J_all[k / num_nodes], 2) - 4 * pow(vals_J__all[k / num_nodes], 2) + pow(vals_J_n_all[k / num_nodes], 2)) / val_dt[0] + beta * derivatives[k / num_nodes];
+			max_Dh = std::max(abs(Dh), max_Dh);
+		}
```

Comment:
To be honest, I am not happy with how the code is written in general. The ordering of the numbers and  inside the various arrays is not very clear to me. Why are there indices like `k / num_nodes` or `k % num_nodes`. These seems very prone to errors.